src/main/java/uk/ac/ebi/atlas/experimentpage/markergenes/HighchartsHeatmapAdapter.java

```diff
@@ -54,14 +58,62 @@ public HighchartsHeatmapAdapter(BioEntityPropertyDao bioEntityPropertyDao) {
+                                .put("pValue", markerGene.pValue())
+                                .build())
+                .collect(toImmutableList());
+    }
+
+    /**
+     * Given a list of marker genes, this method returns a Highcharts data series object
+     * (https://api.highcharts.com/highcharts/series.heatmap.data), where gene IDs/symbols are
+     * the rows (y values), and the cell types are the columns (x values).
+     * The cells contain the median average expression of the gene in the cell group.
+     * The rows of the heatmap are ordered by the cell type, i.e. genes for celltype 1, 2, etc.
+     * If there are no marker genes for a cell group, then no rows will be present in the data.
+     */
+    public ImmutableList<ImmutableMap<String, Object>> getCellTypeMarkerGeneHeatmapData(Collection<CellTypeMarkerGene> markerGenes) {
+        // Whether the comparison by p-value should or shouldn’t be reversed depends on the yAxis.reversed property in
+        // the heatmap component. In our case it’s set to true, so lower p-value is displayed at the top without
+        // reversing the comparator
+        var sortedMarkerGenes = markerGenes.stream()
+                .parallel()
+                .sorted(comparing(CellTypeMarkerGene::cellGroupValueWhereMarker).thenComparing(CellTypeMarkerGene::pValue))
+                .collect(toImmutableList());
+
+        var rows =
+                sortedMarkerGenes.stream()
+                        .map(MARKER_GENE_ID_TO_CELL_GROUP_VALUE_WHERE_MARKER)
+                        .distinct()
+                        .collect(toImmutableList());
+        var columns =
+                sortedMarkerGenes.stream()
+                        .map(CellTypeMarkerGene::cellGroupValue)
+                        .distinct()
+                        .collect(toImmutableList());
+
+        var symbolsForGeneIds =
+                bioEntityPropertyDao.getSymbolsForGeneIds(
+                        sortedMarkerGenes.stream().map(CellTypeMarkerGene::geneId).collect(toImmutableSet()));
+
+        return sortedMarkerGenes.stream()
+                .map(markerGene ->
+                        ImmutableMap.<String, Object>builder()
+                                // To get x co ordinates- extract all distinct cell types as a List(columns in our case)
+                                // and get index of each cell type
+                                .put("x", columns.indexOf(markerGene.cellGroupValue()))
+                                .put("y", rows.indexOf(MARKER_GENE_ID_TO_CELL_GROUP_VALUE_WHERE_MARKER.apply(markerGene)))
+                                .put("geneName", symbolsForGeneIds.getOrDefault(markerGene.geneId(), markerGene.geneId()))
+                                .put("value", markerGene.medianExpression())
+                                .put("cellTypeValue", markerGene.cellGroupValue())
+                                .put("cellTypeValueWhereMarker", markerGene.cellGroupValueWhereMarker())
```

Comment:
Maybe a time will come when we have the same method for both the clusters heatmap and the cell type heatmap, but for now this will do.